### PR TITLE
Suppress cascading diagnostic in default value for parameter

### PR DESCRIPTION
**Pull request by jcouv**
Title says it all ;-)

Fixes https://github.com/dotnet/roslyn/issues/19013 and https://github.com/dotnet/roslyn/issues/16355

### CODE REVIEW

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for review. Small diagnostics fix for 15.6. Thanks

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs [hunk 1]
```diff
@@ -32,6 +33,27 @@ static void Main()
...
+class C
+{
+    async Task M(CancellationToken t = default) { await Task.Delay(0); }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source);
+            comp.VerifyDiagnostics(
+                // (7,40): error CS8107: Feature 'default literal' is not available in C# 7. Please use language version 7.1 or greater.
+                //     async Task M(CancellationToken t = default) { await Task.Delay(0); }
+                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion7, "default").WithArguments("default literal", "7.1").WithLocation(7, 40)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs]: > Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion7, "default").WithArguments("default literal", "7.1").WithLocation(7, 40) [](start = 15, length = 127)

Given the nature of the fix, I believe we might still have a problem with ```default```. In particular, the bound node should not be marked as bad in this scenario. We should still have constant value for the parameter's default.

Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs]: @AlekseyTs When checking value on "default" (which has a parser error), it gets transformed to a bad expression by the code below.

```C#
        internal bool CheckValueKind(BoundExpression expr, BindValueKind valueKind, DiagnosticBag diagnostics)
        {
            if (expr.HasAnyErrors)
            {
                return false;
            }
            ...
        }

        private BoundExpression CheckValue(BoundExpression expr, BindValueKind valueKind, DiagnosticBag diagnostics)
        {
            ...
            if (!hasResolutionErrors && CheckValueKind(expr, valueKind, diagnostics) ||
                expr.HasAnyErrors && valueKind == BindValueKind.RValueOrMethodGroup)
            {
                return expr;
            }

            var resultKind = (valueKind == BindValueKind.RValue || valueKind == BindValueKind.RValueOrMethodGroup) ?
                LookupResultKind.NotAValue :
                LookupResultKind.NotAVariable;

            return ToBadExpression(expr, resultKind);
        }
```

**Comment by AlekseyTs** [comment]
It sounds like we should change implementation of HasAnyErrors helper to ignore any errors that a caused by language version mismatch.

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs]: > if (expr.HasAnyErrors)

It sounds like we should change implementation of HasAnyErrors helper to ignore any errors that a caused by language version mismatch.

Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedDefaultTests.cs]: Filed https://github.com/dotnet/roslyn/issues/20120

**Pull request merged**